python/file_generator.py

```diff
@@ -0,0 +1,377 @@
+
+		#Instantiates the other classes. 
+		
+		#lar_gen is responsible for generating data according to the values
+		# in the schema.
+		self.lar_gen = lar_generator.lar_gen(self.lar_schema_df, 
+			self.ts_schema_df, counties=self.counties, tracts=self.tracts)
+
+		#lar_constrains is responsible for modifying generated data so that 
+		#the resulting file passes syntax and validity edits.
+		self.lar_const = lar_constraints.lar_constraints(counties=self.counties, 
+			tracts=self.tracts) 
+
+		#lar_validator checks a dataframe and returns a JSON with 
+		#edit pass/fail results. 
+		self.lar_validator = rules_engine(lar_schema=self.lar_schema_df, 
+			ts_schema=self.ts_schema_df, cbsa_data=self.cbsas)
+					#tracts=tracts, counties=counties, small_counties=small_counties) 
+
+		#Stores the number of rows in the test file
+		self.file_length = self.data_map["file_length"]["value"] 
+
+		#Stores the LEI for the test file. 
+		self.lei = self.data_map["lei"]["value"]
+
+	def get_const_list(self):
+		
+		"""Creates a list of constraints from the functions in 
+		the lar_constraints object that apply syntax and validity edits."""
+		
+		#Creates an empty list to store constraint function names.
+		constraints = [] 
+
+		#Takes each function pertaining to a syntax or validity edit
+		#in the lar_constraints class, and adds it to the empty constraints.  
+		for func in dir(self.lar_const):
+			if func[:1] in ("s", "v") and func[1:4].isdigit()==True:
+				constraints.append(func)
+
+		#Returns the list of constraints as strings. 
+		return constraints
+
+	def apply_constraint(self, row, func):
+		"""Applies a constraint (func) to a generated row
+		of LAR data and returns a new row in a dictionary format."""
+		
+		#Copies the row. 
+		row_start = row.copy()
+		
+		#Uses getattr to apply the constraint in the lar_constraints class. 
+		row = getattr(self.lar_const, func)(row)
+		
+		#Logs the changes in the intial row after the constraints
+		#have been applied. 
+		diff_1, diff_2 = self.get_diff(row, row_start)
+		if len(diff_1) > 0:
+			logging.info(str(func))
+			logging.info(diff_1)
+			logging.info(diff_2)
+		return row
+
+	def get_diff(self, row, row_base):
+		"""Checks the difference between an initial row and the 
+		row after constraints are applied"""
+		
+		#Convert initial row to set. 
+		initial_row = set(row_base.items()) 
+
+		#Convert constrained row to set. 
+		changed_row = set(row.items()) 
+
+		#Subtract row sets to show changes from constraint functions. 
+		diff_1 = (changed_row - initial_row) 
+		diff_2 = (initial_row - changed_row)
+		return diff_1, diff_2
+
+	def constraints_loop(self, constraints=[], row=None, row_base=None):
+		"""Applies the list of constraints 
+		generated to each row"""
+
+		for const in constraints:
+			row = self.apply_constraint(row, const)
+			diff = self.get_diff(row, row_base)
+		return row
+
+	def validation(self, row, ts_row):
+		"""
+		Applies the rules engine logic 
```

Comment:
Let's specify that it applies the syntax and validity rules engine logic. Unless I'm missing that it also does quality.

Reply:
Sure! This change has been implemented.